apis/metal3.io/v1alpha1/baremetalhost_validation.go

```diff
@@ -0,0 +1,53 @@
+package v1alpha1
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/util/errors"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// log is for logging in this package.
+var log = logf.Log.WithName("baremetalhost-validation")
+
+func (host *BareMetalHost) Validate(old *BareMetalHost) error {
+	log.Info("validate", "name", host.Name)
+	var errs []error
+
+	if err := validateRAID(host.Spec.RAID); err != nil {
+		errs = append(errs, err)
+	}
+
+	if old != nil {
```

Comment:
Wondering if it's really required to flatten create / update validations now. Looks like the code would be simpler if we'll keep them separated

Reply:
In conceptual aspect, update validations also cover create validations. In any cases, all create validations will be executed for also update. I mean even if we flatten this code block, eventually to prevent duplications we are going to conflate them again.

Reply:
For a given resource, the create/delete handler is invoked only once, while the update many. So, once the creation step has been validated, it's not clear why such validation should be repeated also for the update _always_. Of course, there could be validation rules that are meaningful for both the create and update handlers, but there could be other that are valid only in just one of them. For example in the current code the BMC update rule makes sense only during an update.

Reply:
I agree that basically all of the validations in ValidateCreate() are also going to be required in ValidateUpdate() (the only exceptions are stuff that is immutable like the resource name... actually it's just that I think). But I agree with Andrea that there doesn't seem much point taking them from two separate functions, squashing them into one, and then splitting them again based on whether the argument is nil. Like, you could just call ValidateCreate() at the start of ValidateUpdate() and save a lot of boilerplate?

Reply:
Looking deeper, it might make sense to have a common function that returns `[]error` so that both types of error can be combined in a single list, and that's fine too.